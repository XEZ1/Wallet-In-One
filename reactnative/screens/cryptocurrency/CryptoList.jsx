import React, { useEffect } from "react";
import {
  TouchableOpacity,
  SafeAreaView,
  ScrollView,
  StyleSheet,
  Text,
  View,
  Alert,
} from "react-native";
import useCryptoWallet from "../crypto_wallet/useCryptoWallet";
import useCryptoExchange from "../cryptoExchanges/useCryptoExchange";
import CryptoListWalletItem from "../crypto_wallet/CryptoListWalletItem";
import ExchangeAsset from "../cryptoExchanges/ExchangeAsset";
import { useTheme } from '../../src/theme/ThemeProvider'
import { Ionicons } from '@expo/vector-icons';
import * as SecureStore from "expo-secure-store";
import { api_url } from '../../authentication';


export default function CryptoList(props) {
  const { wallets, listWallets, connectWallet, removeWallet } = useCryptoWallet();
  const { exchanges, fetchExchanges, removeExchange } = useCryptoExchange();
  const {dark, colors, setScheme} = useTheme();

  const styles = StyleSheet.create({
    cryptoWalletTitle: {
      fontWeight: "900",
      fontSize: 40,
      alignSelf: "center",
      paddingVertical: 10,
    },
    cryptoWalletSubtitle: {
      fontWeight: "900",
      fontSize: 30,
      alignSelf: "center",
    },

    walletList: {
      marginHorizontal: 10,
      marginBottom: 30,
    },
    button: {
      padding: 10,
      borderRadius: 10,
      marginVertical: 10,
      backgroundColor: colors.primary,
      alignSelf: 'center',
    },
    buttonText: {
      textAlign: 'center',
      fontWeight: 'bold',
      color: colors.text,
    },
    refreshButton: {
      position: 'absolute',
      top: 0,
      right: 10,
      backgroundColor: colors.primary,
      borderRadius: 30,
      padding: 10,
      marginTop: 10
    },
  });

  useEffect(() => {
    listWallets();
    fetchExchanges();
  }, []);

  const handleSubmit = async () => {
    
    try {
      const response = await fetch(api_url + '/crypto-exchanges/update', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Token ${await SecureStore.getItemAsync("token")}`,
        },
        body: JSON.stringify({ }),
      });
      const data = await response.json();
      const statusCode = response.status;
      if (statusCode == 200) {
        Alert.alert('Success', 'updated account data successfully!');
      } else {
        Alert.alert('Error', data["error"]);
      }
    } catch (error) {
      console.error(error);
      Alert.alert('Error', 'An error occurred while updating account info.');
    }
  };

  return (
    <SafeAreaView style={{ flex: 1, backgroundColor: colors.background }}>
      <ScrollView>
        <TouchableOpacity onPress={handleSubmit} style={styles.refreshButton}>
          <Ionicons name="refresh-outline" size={25} color="white" />
        </TouchableOpacity>
        <View style={{ flexDirection: "row", alignItems: "center" }}>
          <View
            style={{
              flex: 1,
              flexDirection: "column",
              justifyContent: "center",
            }}
          >
            <View style={{ marginRight: 20 }}>
              <Text style={[styles.cryptoWalletTitle, {color: colors.text}]}>Cryptocurrency</Text>
            </View>
          </View>
        </View>

        <Text style={[styles.cryptoWalletSubtitle, {color: colors.text, marginTop: 10}]}>Wallets</Text>

        <TouchableOpacity
          style={styles.button}
          onPress={() => props.navigation.navigate("Crypto Wallet Insights")}
        >
          <Text style={styles.buttonText}>Wallet Insights</Text>
        </TouchableOpacity>

        <View style={[styles.walletList]}>
        {
          wallets.map((item) =>
            <CryptoListWalletItem key={item.id} id={item.id} item={item} removeWallet={removeWallet} navigation={props.navigation} />)
        }
        </View>

        <Text style={[styles.cryptoWalletSubtitle, {color: colors.text}]}>Exchanges</Text>
        <View style={[styles.walletList]}>
        {
          exchanges.map((item) =>
            <ExchangeAsset key={item.id} item={item} removeExchange={removeExchange} navigation={props.navigation} />)
        }
        </View>

      </ScrollView>
    </SafeAreaView>
  );
}