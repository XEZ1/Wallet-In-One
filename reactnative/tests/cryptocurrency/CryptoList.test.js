import CryptoList from "../../screens/cryptocurrency/CryptoList";
import {render, screen, act, waitFor, fireEvent} from "@testing-library/react-native";
import cryptoWallets from "../crypto_wallets/fixtures/cryptoWallets.json"

describe('<CryptoList />', () => {

  const params = {
    navigation: {
      navigate: jest.fn(),
      goBack: jest.fn()
    }
  };

  beforeEach(() => {
    global.fetch =  jest.fn( async () => {
      return Promise.resolve({
        status: 200,
      })
    })
  })

  afterEach(() => {
    global.fetch.mockClear();
    delete global.fetch;
  })

  it('snapshot test', async () => {
    const snapshot = render(<CryptoList {...params} />)

    // Test navigation
    await act(async () => {
      fireEvent.press(await screen.getByText('Insights'))
    })

  })

  it('display wallets and exchanges test', async () => {

    global.fetch =  jest.fn( async (api) => {
      const exchangeBalancesPattern = /get_exchange_balances/
      const walletPattern = /crypto_wallets/
      const exchangePattern = /crypto-exchanges/

      if (exchangeBalancesPattern.test(api)) {
        return Promise.resolve({
          status: 200, json: () => ([
            {
              "x": "Binance",
              "y": 0.13,
              "id": 2
            }
          ])
        })
      } else if (walletPattern.test(api)) {
        return Promise.resolve({
          status: 200, json: () => cryptoWallets
        })
      } else if (exchangePattern.test(api)) {
        return Promise.resolve({
          status: 200, json: () => cryptoWallets
        })
      }

    })
    const snapshot = render(<CryptoList {...params} />)

    await act(async () => {

    })

  })

  it('display empty wallets and empty exchanges test', async () => {
    const snapshot = render(<CryptoList {...params} />)

    await act(async () => {
      expect(screen.getByText("There are no wallets to display. Try connect a crypto wallet."))
      expect(screen.getByText("There are no exchanges to display. Try connect a crypto exchange."))
    })

  })

})