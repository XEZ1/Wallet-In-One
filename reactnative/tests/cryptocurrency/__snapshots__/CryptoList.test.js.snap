// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoList /> snapshot test 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "purple",
              "borderRadius": 10,
              "marginHorizontal": 8,
              "marginVertical": 4,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "black",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Insights
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "purple",
              "borderRadius": 10,
              "marginHorizontal": 8,
              "marginVertical": 4,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "black",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Update
          </Text>
        </View>
      </View>
      <Text
        style={
          [
            {
              "alignSelf": "center",
              "fontSize": 30,
              "fontWeight": "900",
            },
            {
              "color": "black",
              "marginTop": 10,
            },
          ]
        }
      >
        Wallets
      </Text>
      <View
        style={
          [
            {
              "marginBottom": 30,
              "marginHorizontal": 10,
            },
          ]
        }
      >
        <Text
          style={
            {
              "color": "black",
            }
          }
        >
          There are no wallets to display. Try connect a crypto wallet.
        </Text>
      </View>
      <Text
        style={
          [
            {
              "alignSelf": "center",
              "fontSize": 30,
              "fontWeight": "900",
            },
            {
              "color": "black",
            },
          ]
        }
      >
        Exchanges
      </Text>
      <View
        style={
          [
            {
              "marginBottom": 30,
              "marginHorizontal": 10,
            },
          ]
        }
      >
        <Text
          style={
            {
              "color": "black",
            }
          }
        >
          There are no exchanges to display. Try connect a crypto exchange.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
